src/libsystemd/sd-event/sd-event.c

```diff
@@ -3893,6 +3893,11 @@ _public_ int sd_event_wait(sd_event *e, uint64_t timeout) {
 
                                 assert(s);
 
+                                if (s->priority > threshold)
+                                        break;
+
+                                min_priority = MIN(min_priority, s->priority);
```

Comment:
this same logic needs to be added a bit further down, where inotify and signal stuff is handled. i.e. we maintain one `signal_data` structure per priority and one `inotify_data` structure per priority. If one of them is triggered we need to take that into account too when determining the minimal priority.

The change should be easy, just do the exact same stuff you do here, but do it on `struct signal_data`'s `priority` field, and the `inotify_data`'s `priority` field.

(for the timer stuff we don't need to do that, since timers are inaccurate anyway)